docs/migration/4_0.md

```diff
@@ -0,0 +1,141 @@
+
+```js
+const app = Module({
+  state: {},
+  providers: {
+    http: HttpProvider({})
+  }
+})
+```
+
+2. The only way to `catch` an error now is at the module level. Errors propagate up to parent modules if not caught. Meaning there is no signal specific error catching or "global". "Global" would be your root level module.
+
+```js
+const app = Module({
+  state: {},
+  catch: [
+    [HttpProviderError, sequences.catchHttpError]
+  ]
+})
+```
+
+[Documentation](/docs/api/module)
+
+## Controller
+The controller takes a module as first argument, and other options as a second argument.
+
+```js
+const controller = Controller(rootModule, {
+  devtools: null
+})
+```
+
+The same goes for server side controller.
+
+[Documentation](/docs/api/controller)
+
+## Provider
+Providers will now have a stronger concept. The concept is being the API between your application and your tools of choice. It can only be defined in one way, an object with methods. This makes things consistent and also enforces the idea that you think of a provider as a "bridge" between your app and the tools your are using.
+
+```js
+const myProvider = Provider({
+  foo() {}
+})
+```
+
+You can point to the existing context using `this.context`. It works this way to simplify with just one API surface and we can do prototypal optimizations under the hood.
+
+```js
+const api = Provider({
+  getUser() {
+    return this.context.http.get('/user')
+  }
+})
+```
+
+All providers defined this way is automatically optimized and wrapped for debugger purposes. If you just want to use a 3rd party tool directly, you can still do that by just attaching it:
+
+```js
+Module({
+  providers: { uuid }
+})
+```
+
+[Documentation](/docs/api/providers)
+
+## Compute
+
+This is just a change to be consistent with the other APIs:
+
+```js
+const myComputed = Compute(state`foo`, (foo) => ())
+```
+
+[Documentation](/docs/api/compute)
+
+## CerebralError
+To more easily create different error types you can now define your errors by extending the CerebralError:
+
+```js
+import { CerebralError } from 'cerebral'
+
+export class AppError extends CerebralError {}
+export class ApiError extends CerebralError {}
+```
+
+[Documentation](/docs/api/error)
+
+## Module state changes
+You can point to the module to make state changes on module executing the signal:
+
+```js
+function myAction ({ module }) {
+  module.set('foo', 'bar')
+}
+```
+You can now use a `module` tag to point to the state of the module running the signal:
+
+```js
+[
+  set(module`foo`, 'bar')
+]
+```
+
+[Documentation tags](/docs/api/tags)
+[Documentation state](/docs/api/state)
+
+## Migration
+
+1. Change controller configuration to take in a top level module
+2. Wrap all modules in Module constructor
+3. Change configuration of providers to use object
+4. Change any custom providers to use the Provider factory
+5. Rename use of `compute` to `Compute`
+5. Any error catching now needs to be at module level without `new Map(...)`
```

Comment:
Maybe add to extend errors from `CerebralError`.
BTW you're using 2 times paragraph **5.** here.